### PR TITLE
cross compile scala 3

### PR DESCRIPTION
**Pull request by i10416**
complete #468 

This PR supports cross build of all modules in Scala 3. 

The most changes are introduced with the aim of hiding shapeless from shared files in core modules and filling the gap of implicit conversions and implicit defs/vals between scala 2.x and 3.x.

In addition, other changes address Scala 3 compiler/linter warnings.

### CODE REVIEW

**Comment by i10416** [comment]
depends on https://github.com/alonsodomin/cron4s/pull/506

**Comment by i10416** [comment]
cron4s compiles in Scala 3!

File: build.sbt [hunk 1]
```diff
@@ -85,27 +89,28 @@ lazy val commonJvmSettings = Seq(
...
-    val tagOrHash = {
-      if (isSnapshot.value)
-        sys.process.Process("git rev-parse HEAD").lineStream_!.head
-      else version.value
-    }
-    val a = (LocalRootProject / baseDirectory).value.toURI.toString
-    val g = "https://raw.githubusercontent.com/alonsodomin/cron4s/" + tagOrHash
-    s"-P:scalajs:mapSourceURI:$a->$g/"
-  },
+  scalacOptions ++= (if (scalaVersion.value.startsWith("2.")) Seq {
```
Comment (i10416) [inline build.sbt:92]: this causes bad option error in scala 3 scala.js:  https://github.com/alonsodomin/cron4s/actions/runs/5435520680

File: modules/testkit/shared/src/main/scala/cron4s/testkit/gen/ArbitraryBetweenNode.scala [hunk 2]
```diff
@@ -24,10 +25,20 @@ import org.scalacheck.Arbitrary
   * Created by alonsodomin on 28/08/2016.
   */
 trait ArbitraryBetweenNode extends NodeGenerators {
-  implicit lazy val arbitraryBetweenSecond     = Arbitrary(betweenGen[Second])
-  implicit lazy val arbitraryBetweenMinute     = Arbitrary(betweenGen[Minute])
-  implicit lazy val arbitraryBetweenHour       = Arbitrary(betweenGen[Hour])
-  implicit lazy val arbitraryBetweenDayOfMonth = Arbitrary(betweenGen[DayOfMonth])
-  implicit lazy val arbitraryBetweenMonth      = Arbitrary(betweenGen[Month])
-  implicit lazy val arbitraryBetweenDayOfWeek  = Arbitrary(betweenGen[DayOfWeek])
+  implicit lazy val arbitraryBetweenSecond: Arbitrary[BetweenNode[Second]] = Arbitrary(
```
Comment (i10416) [inline modules/testkit/shared/src/main/scala/cron4s/testkit/gen/ArbitraryBetweenNode.scala:28]: explicit types for implicit vals/defs

File: modules/testkit/shared/src/main/scala/cron4s/testkit/gen/ArbitraryConstNode.scala [hunk 3]
```diff
@@ -17,17 +17,22 @@
...
 
 import cron4s.CronField._
+import cron4s.expr.ConstNode
 
 import org.scalacheck.Arbitrary
 
 /**
   * Created by alonsodomin on 28/08/2016.
   */
 trait ArbitraryConstNode extends NodeGenerators {
```
Comment (i10416) [inline modules/testkit/shared/src/main/scala/cron4s/testkit/gen/ArbitraryConstNode.scala:27]: explicit types for implicit vals/defs

File: modules/testkit/shared/src/main/scala/cron4s/testkit/gen/ArbitraryCronUnits.scala [hunk 4]
```diff
@@ -25,10 +25,22 @@ import org.scalacheck._
   * Created by alonsodomin on 28/08/2016.
   */
 trait ArbitraryCronUnits {
-  implicit lazy val arbitrarySecondsUnit     = Arbitrary(Gen.const(CronUnit[Second]))
-  implicit lazy val arbitraryMinutesUnit     = Arbitrary(Gen.const(CronUnit[Minute]))
-  implicit lazy val arbitraryHoursUnit       = Arbitrary(Gen.const(CronUnit[Hour]))
-  implicit lazy val arbitraryDaysOfMonthUnit = Arbitrary(Gen.const(CronUnit[DayOfMonth]))
-  implicit lazy val arbitraryMonthsUnit      = Arbitrary(Gen.const(CronUnit[Month]))
-  implicit lazy val arbitraryDaysOfWeekUnit  = Arbitrary(Gen.const(CronUnit[DayOfWeek]))
+  implicit lazy val arbitrarySecondsUnit: Arbitrary[CronUnit[Second]] = Arbitrary(
```
Comment (i10416) [inline modules/testkit/shared/src/main/scala/cron4s/testkit/gen/ArbitraryCronUnits.scala:28]: explicit types for implicit vals/defs

File: project/Dependencies.scala [hunk 5]
```diff
@@ -33,10 +33,12 @@ object Dependencies {
...
   lazy val core = Def.settings(
     libraryDependencies ++= Seq(
-      "com.chuusai"            %%% "shapeless"                % version.shapeless,
       "org.typelevel"          %%% "cats-core"                % version.cats.main,
       "org.scala-lang.modules" %%% "scala-parser-combinators" % version.parserc
-    )
+    ),
+    libraryDependencies ++= (if (scalaVersion.value.startsWith("2."))
+                               Seq("com.chuusai" %%% "shapeless" % version.shapeless)
+                             else Seq.empty)
```
Comment (i10416) [inline project/Dependencies.scala:41]: shapeless is no longer necessary in Scala 3.

File: modules/testkit/shared/src/main/scala-3/cron4s/testkit/gen/CronGenerators.scala [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+ * limitations under the License.
+ */
+
+package cron4s.testkit.gen
+
+import cron4s.CronField
+import cron4s.expr._
+
+import org.scalacheck._
+
```
Comment (i10416) [inline modules/testkit/shared/src/main/scala-3/cron4s/testkit/gen/CronGenerators.scala:23]: This file is almost the same as https://github.com/alonsodomin/cron4s/pull/504/files#diff-fe082e93829b26c97f89813516bdfd1efa4b2f6cc1d26d5dcd7aa2dcec929e03.

The differences are
- explicit type annotations for implicit vals/defs
- shapeless import removal

**Comment by i10416** [comment]
There are around 1700 lines of changes, but changes in test and testkit modules are just explicit type annotations and imports for implicit conversions. Therefore, I think whole changes in this PR are safe to some extent.

**Comment by i10416** [comment]
@alonsodomin Hi, all cron4s modules compile in Scala 3 now. Please check when you have free time.

**Review by alonsodomin — APPROVED** [review]
Thanks a lot for the hard work, I truly believe this was some tedious effort. I'll publish a new version of the library ready for Scala 3 after merging this. Let me know if you have any outstanding things you would like to include before I do that.

**Pull request merged**